src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -80,8 +81,12 @@ void ZooKeeper::init(ZooKeeperArgs args_)
                 if (secure)
                     host_string.erase(0, strlen("secure://"));
 
-                LOG_TEST(log, "Adding ZooKeeper host {} ({})", host_string, Poco::Net::SocketAddress{host_string}.toString());
-                nodes.emplace_back(Coordination::ZooKeeper::Node{Poco::Net::SocketAddress{host_string}, secure});
+                /// We want to resolve all hosts without DNS cache for keeper connection.
+                Coordination::DNSResolver::instance().removeHostFromCache(host_string);
```

Comment:
The errors in integration tests were because docker-compose drops the DNS before the containers actually stopped. So in tests, when stopping a ZK cluster, all nodes are still accessible by IP for a brief period. And since `Coordination::DNSResolver` has a cache in it, some queries in tests that were expected to fail actually completed successfully. 
That can be easily fixed on the side of the test (just wait for cluster termination more precisely). But instead, I decided to remove unnecessary cache from the keeper at all, to avoid potential problems with it.